Src/LiquidProjections/Dispatcher.cs

```diff
@@ -36,44 +39,98 @@ public Dispatcher(CreateSubscription createSubscription)
+        /// behavior is to dispose of the subscription.
+        /// </summary>
+        public HandleException ExceptionHandler { get; set; } = (e, info) => Task.FromResult(ExceptionResolution.Abort);
+
+        private async Task HandleTransactions(IReadOnlyList<Transaction> transactions, Func<IReadOnlyList<Transaction>, SubscriptionInfo, Task> handler, SubscriptionInfo info)
         {
-            try
-            {
-                await handler(transactions, info);
-            }
-            catch (Exception exception)
+            await ExecuteWithPolicy(info, () => handler(transactions, info), abort: exception =>
             {
                 LogProvider.GetLogger(typeof(Dispatcher)).FatalException(
                     "Projector exception was not handled. Event subscription has been cancelled.",
                     exception);
 
                 info.Subscription?.Dispose();
-            }
+            });
         }
 
         private async Task HandleUnknownCheckpoint(SubscriptionInfo info, Func<IReadOnlyList<Transaction>, SubscriptionInfo, Task> handler, SubscriptionOptions options)
         {
             if (options.RestartWhenAhead)
             {
-                try
-                {
-                    info.Subscription?.Dispose();
+                info.Subscription?.Dispose();
 
+                await ExecuteWithPolicy(info, async () =>
+                {
                     await options.BeforeRestarting();
 
                     Subscribe(null, handler, options);
-                }
-                catch (Exception exception)
+                }, abort: exception =>
                 {
                     LogProvider.GetLogger(typeof(Dispatcher)).FatalException(
                         "Failed to restart the projector.",
                         exception);
+                }, ignore: () => Subscribe(null, handler, options));
+            }
+        }
+
+        private async Task ExecuteWithPolicy(SubscriptionInfo info, Func<Task> action, Action<Exception> abort, Action ignore = null)
+        {
+            bool retry = true;
+            do
+            {
+                try
+                {
+                    await action();
+                    retry = false;
+                }
+                catch (Exception exception)
+                {
+                    ExceptionResolution resolution = await ExceptionHandler(exception, info);
+                    switch (resolution)
+                    {
+                        case ExceptionResolution.Abort:
+                            abort(exception);
+                            retry = false;
+                            break;
+
+                        case ExceptionResolution.Retry:
+                            break;
+
+                        case ExceptionResolution.Ignore:
+                            retry = false;
+                            ignore?.Invoke();

```

Comment:
Should we catch and log exceptions thrown from the `abort` and `ignore` delegates?